executable-spec/src/Cardano/Ledger/Spec/STS/Chain/Transactions.hs

```diff
@@ -0,0 +1,160 @@
+  deriving (Eq, Show)
+
+
+instance HeapWords Transaction where
+  heapWords (Dummy dummyTx) = heapWords1 dummyTx
+  heapWords (Ideation ideationTx) = heapWords1 ideationTx
+
+
+instance STS TRANSACTIONS where
+
+  type Environment TRANSACTIONS = Env
+
+  type State TRANSACTIONS = St
+
+  type Signal TRANSACTIONS = [Transaction]
+
+  data PredicateFailure TRANSACTIONS = TransactionFailure (PredicateFailure TRANSACTION)
+    deriving (Eq, Show)
+
+  initialRules = []
+
+  transitionRules = [
+    do
+      TRC (env, st, txs) <- judgmentContext
+      case txs of
+        [] -> pure $! st
+        (tx:txs') -> do
+          st' <- trans @TRANSACTION $ TRC (env, st, tx)
+          trans @TRANSACTIONS $ TRC (env, st', txs')
+    ]
+
+
+instance Embed TRANSACTION TRANSACTIONS where
+  wrapFailed = TransactionFailure
+
+
+data TRANSACTION
+
+
+instance STS TRANSACTION where
+
+  type Environment TRANSACTION = Environment TRANSACTIONS
+
+  type State TRANSACTION = State TRANSACTIONS
+
+  type Signal TRANSACTION = Transaction
+
+  data PredicateFailure TRANSACTION = IdeationFailure (PredicateFailure IDEATION)
+    deriving (Eq, Show)
+
+  initialRules = []
+
+  transitionRules = [
+    do
+      TRC ( Env { currentSlot, participants }
+          , st@St { dummySt, ideationSt }
+          , tx
+          ) <- judgmentContext
+      case tx of
+        Dummy dtx -> do
+          dummySt' <- trans @Dummy.TRANSACTION $ TRC(currentSlot, dummySt, dtx)
+          pure st { dummySt = dummySt' }
+        Ideation itx -> do
+          ideationSt' <- trans @IDEATION $ TRC ( participants, ideationSt, itx )
+          pure st { ideationSt = ideationSt' }
+    ]
+
+
+instance Embed Dummy.TRANSACTION TRANSACTION where
+  wrapFailed = error "Dummy.TRANSACTION shouldn't fail"
+
+
+instance Embed IDEATION TRANSACTION where
+  wrapFailed = IdeationFailure
+
+
+-- | Generate a list of 'Transaction's that fit in the given maximum size.
+transactionsGen
+  :: WordCount
+  -> Environment TRANSACTIONS
+  -> State TRANSACTIONS
+  -> Gen [Transaction]
+transactionsGen maximumSize env st
+  =   fitTransactions . traceSignals OldestFirst
+  <$> genTrace @TRANSACTION 30 env st transactionGen
+
+  where
+    -- Fit the transactions that fit in the given maximum block size.
+    fitTransactions :: [Transaction] -> [Transaction]
+    fitTransactions txs = zip txs (tail sizes)
+                          -- We subtract to account for the block constructor
+                          -- and the 'Word64' value of the slot.
+                        & takeWhile ((< maximumSize - 5) . snd)
+                        & fmap fst
+      where
+        -- We compute the cumulative sum of the transaction sizes. We add 3 to
+        -- account for the list constructor.
+        sizes :: [WordCount]
+        sizes = scanl (\acc tx -> acc + size tx + 3) 0 txs
+
+    transactionGen _ (Env { participants }) (St { ideationSt }) =
+      Gen.frequency [ (9, Dummy <$> Dummy.genTransaction)
```

Comment:
With these frequencies we can have empty Blocks generated (i.e., with no transactions at all). Do we want this?
We also generate slot numbers that are not consecutive. Therefore we have two "sources" for generating empty blocks.